### ISSUE TITLE
Tool built for Node in ESM errors out with: Cannot find module 'is-core-module/package.json'

### ISSUE THREAD

**Issue by kaiyoma** [evanw/esbuild#3102]
I have a large monorepo where one of the packages builds a tool that's used by the rest of the repo. Our repo is currently CJS everywhere, but I'm trying to migrate to ESM. For this tool package, I've done the usual migration steps (set `type` in package.json, switched `main` to `exports`, and set `format` to `esm` for esbuild). The tool seems to build correctly, but when I go to run it, I get this:

```
$ node packages/cv-tools/dist/index.js
node:internal/modules/cjs/loader:1021
  const err = new Error(message);
              ^

Error: Cannot find module 'is-core-module/package.json'
Require stack:
- ...\packages\cv-tools\dist\index.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:1021:15)
    at Function.resolve (node:internal/modules/cjs/helpers:114:19)
    at file:///.../packages/cv-tools/dist/index.js:33:52209
    at file:///.../packages/cv-tools/dist/index.js:4:462
    at file:///.../packages/cv-tools/dist/index.js:33:55403
    at file:///.../packages/cv-tools/dist/index.js:4:462
    at file:///.../packages/cv-tools/dist/index.js:34:800
    at file:///.../packages/cv-tools/dist/index.js:4:462
    at file:///.../packages/cv-tools/dist/index.js:34:9489
    at file:///.../packages/cv-tools/dist/index.js:4:462 {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '...\\packages\\cv-tools\\dist\\index.js'
  ]
}
```

What does this mean? I've never heard of `is-core-module` and we don't use that dependency anywhere in our source code.

Here is our esbuild build file:

```
import { build } from 'esbuild';

build({
  banner: {
    js:
      // This fixes `Error: Dynamic require of <foo> is not supported` in built binaries
      "import { createRequire } from 'module';\n" +
      'const require = createRequire(import.meta.url);\n',
  },
  bundle: true,
  entryPoints: ['src/index.ts'],
  format: 'esm',
  minify: true,
  outdir: 'dist',
  platform: 'node',
  target: 'node16',
}).catch(() => process.exit(1));
```

**Comment by evanw** [comment]
I'm guessing that some package is importing `is-core-module` dynamically, which is not something that esbuild can bundle. This is something you have to debug for yourself. Some ideas:

1. `grep is-core-module -r .` to find the culprit
2. Look in the unminified output file to see what input file is importing `is-core-module` (esbuild emits a comment with the file name at the start of each input file in the bundle)

I'm closing this issue because this is not a problem with esbuild.

**Comment by kaiyoma** [comment]
Ah, this was very helpful, thanks! I'll add some notes here for posterity. I looked in the unminified output and used the comments (which are very helpful!) to determine that it was the `resolve` package which is doing the dynamic import. I then did a `pnpm why` to figure out where `resolve` was coming from, and traced it to [license-checker](https://www.npmjs.com/package/license-checker). That dependency is 4 years old, so it doesn't surprise me that it isn't following sound modern practices. Looks like I'll be looking for a replacement license checker now. Thanks for your help Evan!
